Adds the possibility for GLPI users to discover features with short tutorials

ccailly opened the pull request:
<!--

Dear GLPI user.

BEFORE SUBMITTING YOUR ISSUE, please make sure to read and follow these steps:

* We don't support community plugins. Contact directly their authors, or use the community forum : http://forum.glpi-project.org.
* For feature requests or enhancements, use the suggest dedicated site (http://suggest.glpi-project.org). We check it very often.
* We prefer to keep this tracker in ENGLISH. If you want support in your language, the community forum (http://forum.glpi-project.org) is the best place.
* Please use the below template.

For more information, please check contributing guide:
https://github.com/glpi-project/glpi/blob/main/CONTRIBUTING.md

The GLPI team.
-->


| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 

Adds a discovery system for GLPI functionalities.

![image](https://github.com/glpi-project/glpi/assets/42278610/0833cd93-9ec3-4528-806d-35b7dc2e1502)
![image](https://github.com/glpi-project/glpi/assets/42278610/9863ca05-63d7-48f7-a2d7-7872720db701)
![image](https://github.com/glpi-project/glpi/assets/42278610/7858e501-7a92-48f9-9560-8d768355c736)

orthagh commented on install/migrations/update_10.0.x_to_10.1.0/discover.php:
```diff
@@ -0,0 +1,55 @@
+    $query = "CREATE TABLE `glpi_discover_users` (
+        `id` int {$default_key_sign} NOT NULL AUTO_INCREMENT,
+        `users_id` int {$default_key_sign} NOT NULL DEFAULT '0',
+        `progression` JSON DEFAULT '{}',
```
```
BLOB, TEXT, GEOMETRY or JSON column 'progression' can't have a default value
```

AdrienClairembault reviewed (changes requested):
I think some changes targeting the `/locales` directory were maybe commited by mistake ?

![image](https://github.com/glpi-project/glpi/assets/42734840/dd76d26a-69d2-49bb-8cce-e4d9848ff666)

AdrienClairembault reviewed (changes requested):
Functional review : 

1. The default lesson should not be triggered on the install page
![image](https://github.com/glpi-project/glpi/assets/42734840/ea60795a-becb-495d-8a14-e0b45d4ce171)

2. It would be nice if the last step of the default lesson would show you the menu entry to access all lessons.

3. When a lesson is finished, maybe add a final step that confirm the lesson is over and display a link to go to the lessons page ?

4.  If I keep the main lesson page open in browser tab A, then complete one of the lessons in browser tab B and finally try to run the same lesson from tab A (where its still marked as "not done" because I haven't refreshed the page) I get a blank page.
Maybe in this case allow the user to run the lesson again even if its finished to avoid the error ?

5. I can still scroll on the page when the modals are displayed. 
It feels a bit weird and I'm not sure if it's supposed to work this way.
Is it done on purpose by the intro.js lib and/or is there an option to disable scrolling ?
For example bootstrap modals forbid body scrolling by default:
![image](https://github.com/glpi-project/glpi/assets/42734840/74aa98a9-36d3-4863-97c8-f938fac8a8ea)

6. Lessons are still available for all users but you end up with a blank page when trying to start a lesson if you are not super-admin.
I think we could disable the "My lessons" menu entry for non super-admin users for now ?

7. UI / UX is very good 👍, I still have a few minimal ajustements in mind but I think its best we wait till all the features are implemented before nitpicking on small details ;)

AdrienClairembault commented on src/Discover/Discover_User.php:
```diff
@@ -0,0 +1,177 @@
+
+class Discover_User extends CommonDBTM
+{
+    public static function getTable($classname = null)
```
Don't forget to also change the references to the table in this install/update script

AdrienClairembault commented:
Note: don't forget we need a temporary branch to merge unto as we want to split the development into multiple PRs.

Can somebody with write permission create the branch so we can rebase the PR unto it ?

AdrienClairembault reviewed (changes requested):
Lint is not OK, can you check when you come back @ccailly ?

AdrienClairembault reviewed (changes requested):
It seems lesson content from markdown files are not loaded correctly on my end:

![image](https://github.com/glpi-project/glpi/assets/42734840/dc8a0413-3345-416b-ae40-6fe716b80775)

Is there maybe a compilation step I'm missing ?

I get a 502 error after completing a lesson:

![image](https://github.com/glpi-project/glpi/assets/42734840/6e0eda48-cab0-430d-a829-00e1c582483f)

The lesson is marked correctly as completed.

AdrienClairembault commented on src/Discover/Discover.php:
```diff
@@ -0,0 +1,117 @@
+     */
+    public static function loadDiscover(): void
+    {
+        // Don't load discover if we are on he update page
```
```suggestion
        // Don't load discover if we are on the update page
```

AdrienClairembault commented on templates/discover/lessons.html.twig:
```diff
@@ -0,0 +1,42 @@
+    <table class="table card-table table-vcenter">
+        <tbody>
+            {% for lesson in lessons %}
+                {% include 'discover/lesson.html.twig' with {'progression': progressions[lesson['id']]} %}
```
We are trying to avoid sharing the whole context to twig templates because it gets hard to maintain over time.
It means you should use the `with_context = false` param and share explicitly the `lesson` object.

Also, something I've learned recently is that we should avoid the `include` tag and use the `include` _function_ instead (as described here: https://twig.symfony.com/doc/3.x/tags/include.html).

AdrienClairembault commented on templates/discover/lessons.js.twig:
```diff
@@ -0,0 +1,113 @@
+                });
+
+                {% if endingSteps %}
+                    setTimeout(function () {
+                        launchIntro({{ endingSteps|map(step => step.intro ? step|merge({intro: step.intro|markdown_to_html}) : step)|json_encode|raw }}, true);
+                    }, 0);
```
Could you add a code comment to explain why we need a 0 second timeout here ?

ccailly commented:
> It seems lesson content from markdown files are not loaded correctly on my end:
> 
> ![image](https://private-user-images.githubusercontent.com/42734840/292878410-dc8a0413-3345-416b-ae40-6fe716b80775.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQyNzU2NDYsIm5iZiI6MTcwNDI3NTM0NiwicGF0aCI6Ii80MjczNDg0MC8yOTI4Nzg0MTAtZGM4YTA0MTMtMzM0NS00MTZiLWFlNDAtNmZlNzE2YjgwNzc1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAzVDA5NDkwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZmMmNhODIxZmJkM2NmNWUyODRhYmIxMGUwZDk3Yzk3MmVkNjQwNmFmNDdjZTFmOWI4OGU5N2RiZTE1MDUwZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OERFSD5M2b5-SDsOIUs8UuRkTp6t92T714fgaM9ks8I)
> 
> Is there maybe a compilation step I'm missing ?
> 
> I get a 502 error after completing a lesson:
> 
> ![image](https://private-user-images.githubusercontent.com/42734840/292878529-6e0eda48-cab0-430d-a829-00e1c582483f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQyNzU2NDYsIm5iZiI6MTcwNDI3NTM0NiwicGF0aCI6Ii80MjczNDg0MC8yOTI4Nzg1MjktNmUwZWRhNDgtY2FiMC00MzBkLWE4MjktMDBlMWM1ODI0ODNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAzVDA5NDkwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzNjliNGE3Yjk1MmM3YTFhYzMyNzhmY2ZiYjI1NzFlNmE0ZWQ3MzM2MDJiYzY5MTgwMWNlNjFiZWI0NjMwNDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mDdYVu9-mEi_jTKopHpQ-kg9EeeNqtnKqv7a42l1l6M)
> 
> The lesson is marked correctly as completed.

This error occurred when lessons were not available in the user's language.
Commit 3d0b9e4 corrects this problem by using English by default.

AdrienClairembault reviewed (approved):
Looks good.

Two known issue that we will handle later:
- How to handle links to GLPI's pages in markdown files
- Fix the "triangle path" issue when the next lesson step is not yet displayed (e.g. dropdown value, test the first lesson to see it live)

ccailly commented:
> Looks good.
> 
> Two known issue that we will handle later:
> 
>     * How to handle links to GLPI's pages in markdown files
> 
>     * Fix the "triangle path" issue when the next lesson step is not yet displayed (e.g. dropdown value, test the first lesson to see it live)

The "triangle path" problem has been solved in this commit 5f32045

AdrienClairembault commented:
Besides pending reviews, could someone also create a dedicated branch in which this could be merged ? (ping @cedric-anne).

cedric-anne commented on templates/discover/lessons.js.twig:
```diff
@@ -0,0 +1,122 @@
+ #}
+
+<script>
+    function launchIntro(steps, isEndingLesson = false) {
```
It would be better to externalize most of the javascript in a `js` file. Indeed, javascript inside Twig files are not processed by lint tasks and cannot be tested with Jest.

orthagh reviewed (commented):
Tests of the days:

- lessons repo was not up to date, error reported in logs:
```
[2024-01-29 14:56:56] glpiphplog.CRITICAL:   *** Uncaught Exception TypeError: array_walk_recursive(): Argument #1 ($array) must be of type array, int given in /var/www/html/glpi/main/src/Discover/Discover_User.php at line 146
```
- After update, it still fails on others lessons on my side, with a log: 
```
[2024-01-29 15:23:18] glpiphplog.WARNING:   *** PHP Warning (2): foreach() argument must be of type array|object, null given in /var/www/html/glpi/main/vendor/twig/twig/src/Extension/CoreExtension.php at line 1695
  Backtrace :
  ...ates/30/30c1f1e67a16ab7ba2732d3d9e17b1db.php:46 twig_array_map()
  vendor/twig/twig/src/Template.php:394              __TwigTemplate_7c863886c16976a0a8246c1a67f306db->doDisplay()
  vendor/twig/twig/src/Template.php:367              Twig\Template->displayWithErrorHandling()
  vendor/twig/twig/src/TemplateWrapper.php:45        Twig\Template->display()
  src/Application/View/TemplateRenderer.php:198      Twig\TemplateWrapper->display()
  src/Discover/Discover.php:75                       Glpi\Application\View\TemplateRenderer->display()
  src/Html.php:1789                                  Glpi\Discover\Discover::loadDiscover()
  front/ticket.php:63                                Html::footer()
  public/index.php:82                                require()
```

Don't know why the last appears, any hints welcomed.

Regarding the config file in lessons folder, I see this kind of path: ` "content" => "file://introduction.md",`. I suggest to move to `file://./sources/introduction.md` to permits to click on links in vscode.

cconard96 commented on src/Discover/Discover.php:
```diff
@@ -0,0 +1,118 @@
+                'steps' => $lesson['steps'],
+                'endingSteps' => isset($lesson['showEndingLesson']) && $lesson['showEndingLesson']
+                    ? $discover_user->getEndingLesson()['steps']
+                    : null,
```
Should be an empty array instead of null to avoid passing null to array_map

cconard96 commented on ajax/discover.php:
```diff
@@ -0,0 +1,57 @@
+    switch ($_POST['action']) {
+        case 'lesson-done':
+            if (!isset($_POST['lesson_id'])) {
+                header('HTTP/1.1 400 Bad Request');
```
```suggestion
                http_response_code(400);
```
I don't think we gain anything by using `header`.

cconard96 commented on ajax/discover.php:
```diff
@@ -0,0 +1,57 @@
+            break;
+    }
+} else {
+    header('HTTP/1.1 400 Bad Request');
```
```suggestion
    http_response_code(400);
```

cconard96 commented on js/discover.js:
```diff
@@ -0,0 +1,119 @@
+/* global introJs */
+
+function launchIntro(endpoint, lessonId, steps, endingSteps = [], isEndingLesson = false) {
+    var intro = introJs();
```
```suggestion
    const intro = introJs();
```

cconard96 commented on resources/Lessons/config.php:
```diff
@@ -0,0 +1,146 @@
+ * ---------------------------------------------------------------------
+ */
+
+return [
```
If the lessons are going to be part of the GLPI repo, the translations would benefit from having a context specified.

cconard96 commented on resources/Lessons/config.php:
```diff
@@ -0,0 +1,146 @@
+                    "element" => ".navbar-brand",
+                    "tooltipPosition" => "right",
+                    "title" => __("Menu"),
+                    "content" => __("Vous pouvez accéder à la page d'accueil en cliquant sur le logo GLPI en haut à gauche."),
```
```suggestion
                    "content" => __("You can access the home page by clicking on the GLPI logo at the top left."),
```

cconard96 commented on resources/Lessons/config.php:
```diff
@@ -0,0 +1,146 @@
+            "id" => "create-ticket",
+            "title" => __("Create a ticket"),
+            "category" => "assistance",
+            "description" => __("Create our first ticket"),
```
```suggestion
            "description" => __("Create your first ticket"),
```

cconard96 commented on resources/Lessons/sources/assistance/create-ticket.md:
```diff
@@ -0,0 +1 @@
+You can now create a ticket by clicking on this button.
```
```suggestion
You can create a ticket now by clicking on this button.
```

cconard96 commented on resources/Lessons/sources/assistance/empty-ticket-page.md:
```diff
@@ -0,0 +1,2 @@
+This is the page with all your tickets.
+You can see the tickets you have created, those to which you are assigned, ...
```
```suggestion
You can see the tickets you have created, those to which you are assigned, etc.
```

cconard96 commented on resources/Lessons/sources/assistance/ticket-dashboard.md:
```diff
@@ -0,0 +1,3 @@
+This section groups together several indicators in the form of a dashboard.
+Indicators can be edited by clicking on the "Modify" button at the top right of the page.
```
```suggestion
Indicators can be edited by clicking on the "Toggle edit mode" button at the top right of the page.
```

cconard96 commented on resources/Lessons/sources/introduction.md:
```diff
@@ -0,0 +1,9 @@
+Hello 👋 and welcome to GLPI! This is the beginning of your adventure into the world of Free and OpenSource ITSM software designed and written by Teclib'. Throughout this wizard, I'll guide you step by step so that you get off to a good start and don't feel lost in this ocean of settings :).
```
```suggestion
Hello 👋 and welcome to GLPI! This is the beginning of your adventure into the world of Free and Open Source ITSM software designed and written by Teclib'. Throughout this wizard, you will be guided step by step through some of the basic GLPI functionality.
```
Lesson text should not refer to itself as a person. I think the tone, besides the first (and maybe second) sentences, needs to be more formal too.

cconard96 commented on templates/layout/parts/page_header.html.twig:
```diff
@@ -33,6 +33,8 @@
 
 {% set anonymous = session('glpiactiveprofile') is null %}
 
+{% set is_super_admin = session('glpiactiveprofile')['id'] in call('Profile::getSuperAdminProfilesId') %}
```
```suggestion
{% set is_super_admin = has_profile_right('config', constant('UPDATE')) %}
```

AdrienClairembault reviewed (changes requested):
Some lessons content can't be fetched if the locates aren't compiled.

![image](https://github.com/glpi-project/glpi/assets/42734840/b980508f-c76e-46ce-a81d-96ff8cc7c73e)

I think in this case we should still be able to display the default (english) content.

Also the compile scripts doesn't seem to work on my side, can you test it again to make sure it is OK ?

ccailly commented:
> Some lessons content can't be fetched if the locates aren't compiled.
> 
> ![image](https://private-user-images.githubusercontent.com/42734840/302587110-b980508f-c76e-46ce-a81d-96ff8cc7c73e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyMjgxMTgsIm5iZiI6MTcwNzIyNzgxOCwicGF0aCI6Ii80MjczNDg0MC8zMDI1ODcxMTAtYjk4MDUwOGYtYzc2ZS00NmNlLWE4MWQtOTZmZjhjYzdjNzNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA2VDEzNTY1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFjM2U3MzkyNzcyNTgyNTU3NDc3ZDZkODk2ZTgzNjBhYmE2MjcxNzMwZmYzZDJjZjQ0YWQ0MWE2OGViOWQ0NjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QzNhCDhsU72OuyPf8FFJxbu7eewQNX-IJysx-JcP9OA)
> 
> I think in this case we should still be able to display the default (english) content.
> 
> Also the compile scripts doesn't seem to work on my side, can you test it again to make sure it is OK ?

I've added a fallback on the sources in case no translation exists.
To run the build script, you need to be in the `resources/Lessons` folder. This script is temporary and will disappear or be modified.

AdrienClairembault commented:
Any updates on this ? Some reviews are still pending.

ccailly committed:
Adds the possibility for GLPi users to discover features with short tutorials

Revert composer.lock unexpected changes

fix: resolve promise immediately if dropdown menu is not animated

fix: assign returned config array to variable

refactor: move launchIntro function in a separate JS file

fix: JS lint

fix: 1 file without header

fix: update lessons repo version

fix: change default value of endingSteps to empty array

refactor: Move lessons in the main GLPI repository

refactor: simplify dropdown-show action data structure

refactor: update of markdown file sources to enable clicking on links in vscode

fix: CI

Apply suggestions from code review

Co-authored-by: Curtis Conard <cconard96@gmail.com>
fix: fallback to the sources if the english translation doesn't exist

ccailly committed:
fix(tests): Fix and add E2E test

ccailly committed:
fix: JS lint

ccailly committed:
refactor(discover): rebase code

ccailly committed:
fix: add missing header

cconard96 commented on js/modules/Discover.js:
```diff
@@ -0,0 +1,141 @@
+        document.body.style.overflow = 'hidden';
+
+        if (this.steps && this.steps[this._currentStep] && this.steps[this._currentStep].actions) {
+            var step = this.steps[this._currentStep];
```
`var` really shouldn't be needed anywhere. `let` and `const` should always be used instead.

cconard96 commented on js/modules/Discover.js:
```diff
@@ -0,0 +1,141 @@
+
+    #handleBeforeChange(intro) {
+        // Disable scroll when intro is running
+        document.body.style.overflow = 'hidden';
```
Better handled by adding and later removing "overflow-hidden" class on the body element.

ccailly committed:
chore: replace var and use class instead of style

cedric-anne commented:
I put this back to draft. It will have to be rebased when this topic will be ready to be finished.
